index.js

```diff
@@ -4,79 +4,153 @@ var React = require('react');
-      //
-      // We check for `== null` to cover both `null` and `undefined`, this is
-      // what React does to check if component is used in uncontrolled mode.
-      this.props.value == null
-    ) {
-      this.recalculateSize();
-    }
-  },
+    displayName: 'TextareaAutosize',
 
-  onChange: function(e) {
-    if (this.props.onChange) {
-      this.props.onChange(e);
-    }
-    if (this.props.value === undefined) {
-      // controlled mode
-      this.recalculateSize();
-    }
-  },
-
-  recalculateSize: function() {
-    var diff;
-    var node = this.getDOMNode();
-
-    if (window.getComputedStyle) {
-      var styles = window.getComputedStyle(node);
-
-      // If the textarea is set to border-box, it's not necessary to
-      // subtract the padding.
-      if (styles.getPropertyValue('box-sizing') === "border-box" ||
-          styles.getPropertyValue('-moz-box-sizing') === "border-box" ||
-          styles.getPropertyValue('-webkit-box-sizing') === "border-box") {
-        diff = 0;
-      } else {
-        diff = (
-          parseInt(styles.getPropertyValue('padding-bottom') || 0, 10) +
-          parseInt(styles.getPropertyValue('padding-top') || 0, 10)
-        );
-      }
-    } else {
-      diff = 0;
-    }
+    propTypes: {
+        minHeight: React.PropTypes.number,
+        maxHeight: React.PropTypes.number
+    },
+
+    /**
+     * This component can work with the ValueLink shortcut provided by React.
+     * When ValueLink is passed, it tries to take over the onChange event in the underlying element
+     * However, we need onChange to calculate the height of the box ourselves
+     * So we store ValueLink locally, remove it from the underlying <textarea>, but proxy all onChange calls to it
+     */
+    valueLink: undefined,
+
+    /**
+     * A cached copy of properties that will be passed to the underlying <textarea>
+     * Essentially passes all own properties, but sanitizes them to remove ValueLink and add the necessary style settings
+     */
+    textareaProps: {},
+
+    getInitialState: function () {
+        return {
+            height:   'auto',
+            overflow: 'hidden'
+        }
+    },
+
+    render: function () {
+        this.textareaProps.style = objectAssign({}, this.textareaProps.style, {
+            height:   this.state.height,
+            overflow: this.state.overflow
+        });
+
+        return React.DOM.textarea(this.textareaProps);
+    },
+
+    componentWillMount: function () {
+        this.textareaProps = this.sanitizeProps(this.props);
+    },
+
+
+    /**
+     * Do the initial calcluation of the height
+     * And set an event listener to update the textarea height when the window is resized
+     */
+    componentDidMount: function () {
+        //do a short timeout so let the browser sort out its paddings etc
+        //otherwise sometimes the browser doesn't not text should be wrapped and missed some cases
+        setTimeout(this.recalculateSize);
```

Comment:
For now I put up the relevant pieces at https://gist.github.com/eugene1g/5dbaa7d35d0c7d5c2c56 as an implementation reference.

Couple disclaimers:
1. The code is structured for ES6 and React 0.13, so you may need to have it transpiled (we use babel across all scripts so that's not an issue here)
2. There are two files - one is React-specific that renders the component, and the other one is DOM-specific that measures how long the text should be. This is because we still have old-school DOM interactions elsewhere in the app, and need to have the measuring logic available elsewhere. But it could be rolled up into the same component
3. The measuring script injects an additional invisible textarea into the DOM for all of its work and styles it accordingly to make sure its height measurements are the same as the real textarea node. This is necessary because adjusting the height on the React dom node directly often ends up messing up the scrolling position on the page.
4. For performance reasons, the DOM measuring util caches all relevant CSS descriptors of the textarea. This works well for us, but if you're doing dynamic manipulation of paddings / line-heights etc then you may need to bust that cache or remove it altogether.